Simplify GTK code

k-takata opened the pull request:
This is a supplemental patch for #3421.
The GTK code still has too many #ifs.
Define some functions and remove old code for GTK1.
E.g. No need to use `gtk_signal_connect()` any more.

This also drops support for Glib 2.3 or earlier.
`G_DEFINE_TYPE` can be used from Glib 2.4.0 (released on 2004).
I don't think we need to support older Glib than that.

k-takata committed:
Simplify the GTK code

This is a supplemental patch for #3421.
The GTK code still has too many #ifs.
Define some functions and remove old code for GTK1.
E.g. No need to use `gtk_signal_connect()` any more.

k-takata committed:
Drop support for Glib 2.3 or earlier

`G_DEFINE_TYPE` can be used from Glib 2.4.0 (released on 2004).
I don't think we need to support older Glib than that.

k-takata commented:
I hope someone try this especially with old GTK2.
@arenieri Could you try this?

@brammool Do you think it's okay to drop support for Glib 2.3 or earlier?

brammool commented:
> I hope someone try this especially with old GTK2.
> @arenieri Could you try this?
> 
> @brammool Do you think it's okay to drop support for Glib 2.3 or earlier?

I don't know.  Perhaps there are some older systems where upgrading is
difficult?

-- 
Time is money.  Especially if you make clocks.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

tgc commented:
The GTK2 support is currently not buildable with gtk < 2.10, patch 8.0.1211 broke it.
Reverting that allows vim to build with even glib 2.2.3/gtk 2.2.4 (RHEL 3) and glib 2.4.7/gtk 2.4.13 (RHEL 4). The most recent build I did was 8.1.369.
I did a (very) brief test and gvim seems to be functioning on both platforms.
Both platforms are long obsolete though so this is probably not something that should hold too much weight.

k-takata commented:
> patch 8.0.1211 broke it.

Aha, that was written by me. It uses `gtk_notebook_set_tab_reorderable()` which was added in GTK 2.10, and it is the most important part of the patch.
No one complained this (before you), and the extened support of RHEL4 has been ended last year, so I don't think we should keep supporting before GTK 2.10.
Keep the code base simpler is important for the future development.

Anyway, does anyone try this PR on RHEL5 or other older environments?

tgc commented:
The build fails on el5 (8.1.401+PR3436):
```
objects/gui_gtk_x11.o: In function `drag_data_received_cb':
/builddir/build/BUILD/vim-8.1.0401/src/gui_gtk_x11.c:2250: undefined reference to `gtk_selection_data_get_format'
objects/gui_gtk_x11.o: In function `drag_handle_text':
/builddir/build/BUILD/vim-8.1.0401/src/gui_gtk_x11.c:2212: undefined reference to `gtk_selection_data_get_data_type'
objects/gui_gtk_x11.o: In function `selection_received_cb':
/builddir/build/BUILD/vim-8.1.0401/src/gui_gtk_x11.c:1421: undefined reference to `gtk_selection_data_get_data_type'
/builddir/build/BUILD/vim-8.1.0401/src/gui_gtk_x11.c:1426: undefined reference to `gtk_selection_data_get_data_type'
collect2: ld returned 1 exit status
link.sh: Linking failed
```

Also I'd forgotten but gtk < 2.4.0 support was broken by 7.4.2042 and the el3 builds I've done are patched to fall back to gdk_display_sync().

brammool commented:
> > patch 8.0.1211 broke it.
> 
> Aha, that was written by me. It uses
> `gtk_notebook_set_tab_reorderable()` which was added in GTK 2.10, and
> it is the most important part of the patch.
> No one complained this (before you), and the extened support of RHEL4
> has been ended last year, so I don't think we should keep supporting
> before GTK 2.10.
> Keep the code base simpler is important for the future development.
> 
> Anyway, does anyone try this PR on RHEL5 or other older environments?

I am wondering, suppose someone on such a RHEL4 system wants to build
the latest Vim, would it be possible to install GTK 2.10 or later?
In my experience (from FreeBSD) there are so many dependencies that this
can be quite hard.  Well, I suppose it's still possible to build a
non-GUI version.

It's been a tradition with Vim to keep supporting older systems for as
long as it's doable.  Support expiration is done for commercial reasons,
it doesn't mean those systems suddenly disappear.

-- 
hundred-and-one symptoms of being an internet addict:
82. AT&T names you Customer of the Month for the third consecutive time.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

k-takata committed:
Fix build error on RHEL5

k-takata committed:
Revert "Drop support for Glib 2.3 or earlier"

This reverts commit 9fb39ed60636e07cec667a4a43cfae59e28e7348.

Someone still may want to build Vim with GTK 2.2.

k-takata commented:
Okay, so the current status (without my PR) is:

* The latest Vim requires GTK 2.10 (RHEL5).
* If the patch 8.0.1211 is reverted, it can be built with GTK 2.4 (RHEL4).
* If the patches 8.0.1211 and 7.4.2042 are reverted, it can be built with GTK 2.2 (RHEL3).

@tgc Thank you for testing. I have updated the PR to fix these errors and revert [my second commit](https://github.com/vim/vim/pull/3436/commits/9fb39ed60636e07cec667a4a43cfae59e28e7348) which requires Glib 2.4.

So the current status with my PR is:

* Vim should be built with GTK 2.10.
* If the patch 8.0.1211 is reverted (with modification), it should be built with GTK 2.4.
* If the patches 8.0.1211 and 7.4.2042 are reverted (with modification), it should be built with GTK 2.2.
* Of course, the code becomes simpler because many #ifs are removed.

> I am wondering, suppose someone on such a RHEL4 system wants to build the latest Vim, would it be possible to install GTK 2.10 or later?

Updating GTK might be very difficult. But my updated PR is be acceptable?

tgc commented:
> I am wondering, suppose someone on such a RHEL4 system wants to build the latest Vim, would it be possible to install GTK 2.10 or later?

Later releases of RHEL 4 already include GTK 2.10 since Red Hat added backports of the RHEL 5 glib/gtk stack to support evolution 2.8.x.
The catch is they were built into a private location so using them requires adding these non-standard locations to the build environment and then ensuring they are used at runtime with vim over the versions in the standard locations.

tgc commented:
> @tgc Thank you for testing. I have updated the PR to fix these errors and revert [my second commit](https://github.com/vim/vim/pull/3436/commits/9fb39ed60636e07cec667a4a43cfae59e28e7348) which requires Glib 2.4.
> 

I can confirm that this PR now works as is on RHEL 5. Additionally I've modified the patches needed for gtk < 2.10 to apply on top of this PR and can confirm that with those applied it also works on RHEL 3 and RHEL 4.

k-takata commented:
Thank you for testing again. I think this PR is ready to be merged.
No requirements are changed with or without this PR, and the GTK code is now really simplified.

k-takata committed:
Remove very old comments for GTK1 code

brammool commented:
> Okay, so the current status (without my PR) is:
> 
> * The latest Vim requires GTK 2.10 (RHEL5).
> * If the patch 8.0.1211 is reverted, it can be built with GTK 2.4 (RHEL4).
> * If the patches 8.0.1211 and 7.4.2042 are reverted, it can be built with GTK 2.2 (RHEL3).
> 
> @tgc Thank you for testing. I have updated the PR to fix these errors and revert [my second commit](https://github.com/vim/vim/pull/3436/commits/9fb39ed60636e07cec667a4a43cfae59e28e7348) which requires Glib 2.4.
> 
> So the current status with my PR is:
> 
> * Vim should be built with GTK 2.10.
> * If the patch 8.0.1211 is reverted (with modification), it should be built with GTK 2.4.
> * If the patches 8.0.1211 and 7.4.2042 are reverted (with modification), it should be built with GTK 2.2.
> * Of course, the code becomes simpler because many #ifs are removed.
> 
> > I am wondering, suppose someone on such a RHEL4 system wants to build the latest Vim, would it be possible to install GTK 2.10 or later?
> 
> Updating GTK might be very difficult. But my updated PR is be acceptable?

How about aiming for RHEL4, so GTK 2.4?  I don't have statistics of how
many of these machines are still running, but RHEL3 seems really old.

We might also consider what versions of GTK actually work well.  I
recall that some of the early 2.x versions weren't really usable.

-- 
ASCII stupid question, get a stupid ANSI.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

k-takata commented:
Somehow, [@brammool's comment in vim_dev](https://groups.google.com/d/msg/vim_dev/FkulKvhN-Tk/wtbxo1ccCgAJ) is not redirected here. So quoting:

> > > @tgc Thank you for testing. I have updated the PR to fix these errors and revert [my second commit](https://github.com/vim/vim/pull/3436/commits/9fb39ed60636e07cec667a4a43cfae59e28e7348) which requires Glib 2.4.
> > >
> >
> > I can confirm that this PR now works as is on RHEL 5. Additionally
> > I've modified the patches needed for gtk < 2.10 to apply on top of
> > this PR and can confirm that with those applied it also works on RHEL
> > 3 and RHEL 4.
> 
> Thanks for checking.  I'll include the (large!) change from Ken, then
> please check and send us your change.

Just reverting the pach 8.0.1211 (with the modification) spoils its functionality on GTK 2.10 or later. So we should use another #ifs for that...
I will create a new PR.

k-takata commented:
I created #3447.

@brammool 
> How about aiming for RHEL4, so GTK 2.4? I don't have statistics of how
> many of these machines are still running, but RHEL3 seems really old.

If we target GTK 2.4, then could you include this patch? https://github.com/vim/vim/pull/3436/commits/9fb39ed60636e07cec667a4a43cfae59e28e7348
(You can get the raw diff by appending `.diff` to the URL:
https://github.com/vim/vim/pull/3436/commits/9fb39ed60636e07cec667a4a43cfae59e28e7348.diff )
This drops the old code for Glib 2.2, but if there are someone still want to use RHEL3, they can easily do it by reverting this (and also 7.4.2042).
